lib/web_ui/lib/src/engine/bitmap_canvas.dart

```diff
@@ -568,30 +579,76 @@ class BitmapCanvas extends EngineCanvas with SaveStackTracking {
     _applyPaint(paint);
     final HtmlImage htmlImage = image;
-    final html.Element imgElement = htmlImage.imgElement.clone(true);
+    final html.Element imgElement = htmlImage.cloneImageElement();
+    _drawImage(imgElement, p);
+    _childOverdraw = true;
+  }
+
+  void _drawImage(html.ImageElement imgElement, ui.Offset p) {
     imgElement.style
-      ..position = 'absolute'
-      ..transform = 'translate(${p.dx}px, ${p.dy}px)';
-    rootElement.append(imgElement);
+      ..position = 'absolute';
+    if (isClipped) {
+      final List<html.Element> clipElements =
+      _clipContent(_clipStack, imgElement, p, currentTransform);
+      for (html.Element clipElement in clipElements) {
+        rootElement.append(clipElement);
+        _children.add(clipElement);
+      }
+    } else {
+      final String cssTransform =
+      matrix4ToCssTransform(transformWithOffset(currentTransform, p));
+      imgElement.style
+        ..transformOrigin = '0 0 0'
+        ..transform = cssTransform;
+      rootElement.append(imgElement);
+      _children.add(imgElement);
+    }
   }
 
   @override
   void drawImageRect(
       ui.Image image, ui.Rect src, ui.Rect dst, ui.PaintData paint) {
-    // TODO(het): Check if the src rect is the entire image, and if so just
-    // append the imgElement and set it's height and width.
     final HtmlImage htmlImage = image;
-    ctx.drawImageScaledFromSource(
-      htmlImage.imgElement,
-      src.left,
-      src.top,
-      src.width,
-      src.height,
-      dst.left,
-      dst.top,
-      dst.width,
-      dst.height,
-    );
+    final bool requiresClipping = src.left != 0 || src.top != 0 ||
+      src.width != image.width || src.height != image.height;
+    if (dst.width == image.width && dst.height == image.height &&
+        !requiresClipping) {
+      drawImage(image, dst.topLeft, paint);
+    } else {
+      _applyPaint(paint);
+      final html.Element imgElement = htmlImage.cloneImageElement();
+      if (requiresClipping) {
+        save();
+        clipRect(dst);
+      }
+      _drawImage(imgElement, ui.Offset(dst.left, dst.top));
+      // To scale set width / height on destination image.
+      // For clipping we need to scale according to
+      // clipped-width/full image width and shift it according to left/top of
+      // source rectangle.
+      double targetWidth = dst.width;
+      double targetHeight = dst.height;
+      if (requiresClipping) {
+        targetWidth *= image.width / src.width;
+        targetHeight *= image.height / src.height;
+      }
+      final html.CssStyleDeclaration imageStyle = imgElement.style;
+      imageStyle
+        ..width = '${targetWidth.toStringAsFixed(2)}px'
+        ..height = '${targetHeight.toStringAsFixed(2)}px';
+      if (requiresClipping) {
+        if (src.width != image.width) {
+          double leftMargin = -src.left * (dst.width / src.width);
+          imageStyle.marginLeft = '${leftMargin.toStringAsFixed(2)}px';
```

Comment:
IICR, margin is a layout-affecting CSS property. Dunno how much it matters here, but it may be slower than a transform, which only affects painting.

Reply:
thx. switched to use offset.